docs/extensions/storage-transformers/sharding/v1.0.rst

```diff
@@ -0,0 +1,293 @@
+  the index to update it.
+
+* ``erase_values(keys)``: For each referenced key, check if all available chunks
+  in a shard are referenced. In this case, the full shard is removed using
+  ``erase_values`` on the underlying store. For all other keys, request the
+  indices from the underlying store using ``get_partial_values``. Update the
+  index using an offset and nbytes of ``2^64 - 1`` to mark missing chunks. The
+  updated index may be written in-place using
+  ``set_partial_values(key_start_values)``, or a larger rewrite of the shard may
+  be done including the index update, but also removing value ranges
+  corresponding to the erased chunks.
+
+* ``erase_prefix()``: If the prefix contains a part of the chunk-grid key, this
+  part is translated to the referenced shard and contained chunks. For affected
+  shards where all contained chunks are erased the prefix is rewritten to the
+  corresponding shard key and the operation is relayed to the underlying store.
+  For all shards where only some chunks are erased the affected chunks are
+  removed by invoking the operation ``erase_values`` on this storage transformer
+  with the respective chunk keys.
+
+* ``list()``: See ``list_prefix`` with the prefix ``/``.
+
+* ``list_prefix(prefix)``: If the prefix contains a part of the chunk-grid key,
+  this part is translated to the referenced shard and contained chunks. Then,
+  ``list_prefix`` is called on the underlying store with the translated prefix.
+  For all listed shards request the indices from the underlying store using
+  ``get_partial_values``. Existing chunks, where the index offset or nbytes are
+  not ``2^64 - 1`` are then listed by their original key.
+
+* ``list_dir(prefix)``: If the prefix contains a part of the chunk-grid key,
+  this part is translated to the referenced shard and contained chunks. Then,
+  ``list_dir`` is called on the underlying store with the translated prefix. For
+  all *retrieved prefixes* (not full keys) with partial shard keys, the
+  corresponding original prefixes covering all possible chunks in the shard are
+  listed. For *retrieved full keys* the indices from the underlying store are
+  requested using ``get_partial_values``. Existing chunks, where the index
+  offset or nbytes are not ``2^64 - 1`` are then listed by their original key.
+
+  .. note::
+
+    Not all listed prefixes must necessarily contain keys, as shard prefixes
+    with partially available chunks return prefixes for all possible chunks
+    without verifying their existence for performance reasons. Listing those
+    prefixes is still safe as some chunks in their corresponding shard exist,
+    but not necessarily in the requested prefix, possibly leading to empty
+    responses. Please note that this only applies to returned prefixes, *not*
+    for full keys referencing storage objects. Returned full keys always reflect
+    the available chunks and are safe to request.
```

Comment:
I read this section like 5 times, but I still can't understand it. 🙃  Consider rewriting it from scratch.

Reply:
Yes, this is a quite confusing part. It is actually proposing a solution to [the problem you also mentioned above](https://github.com/zarr-developers/zarr-specs/pull/152#discussion_r942533323), but on the `list_dir` method:
> Just noting that, if a read of each chunk needs to be done to list a prefix, this simple list operation could get pretty expensive for large arrays.

This holds true here as well. In this case, I directly proposed the same solution I also mentioned above; to simply list all chunks that **might** be in a chunk, not checking the index for ``2^64 - 1`` entries. Does that help to get the idea of the note across?

I think the same strategy should be used both for `list_dir` and `list_prefix`. Before changing the related parts, I'd like to get your and possibly more opinions on those questions:
* Should the default behavior be "exact, but possibly slow" (checking all indices), or rather "fast, but possibly containing directories and keys for chunks which actually don't exist"?
* Should the other version be possible as well, maybe as a runtime configuration?
* Should any of this be part of the spec, or are those implementation details?

Reply:
Note: At the moment nothing in the spec actually uses this operation.  For array operations, only `set` and `get` are used currently.  The only references to list operations are for listing groups, which isn't relevant to sharding.

In general I think we could eliminate all of the descriptions of operations from the spec, since I think they are all basically self-evident from the description of the format itself.  I think only in cases where there is some specific subtle detail that needs to point out should we bother defining operations.

For example, I don't think I would bother to refer to any of the operation descriptions when creating the zarr v3 implementations for TensorStore or Neuroglancer.

Additionally I think we have enough issues to resolve with the actual format that we could save ourselves the added trouble of carefully specifying a description of all of the operations.  We could always add the descriptions later if some implementers think they are valuable.

Reply:
I agree with the general principle of keeping the spec focused more narrowly on the format itself and leaving the "operations" to the implementations. However, what you have written up here is very useful for implementers. So it would be a shame to just throw it out.

Reply:
Thank you very much for the feedback! I agree that the operation descriptions are just examples or guidelines for possible implementations, and I think they might help there. I added a couple of changes in 9b3237c regarding those:
* noted that the operation descriptions are not normative
* added the core meaning of the previous note to the `list_dir` description
* changed the note to emphasize that the operations might even be left out, and could potentially be implemented with the efficient variant of listing all possible chunks.

What do you think about the updated version @rabernat @jbms?

Reply:
@rabernat @jbms Please let me know if your are missing anything here @rabernat @jbms. In the meantime I'll simply assume that this is resolved.